sdk/cosmos/azure-cosmos-spark_3_2-12/src/main/scala/com/azure/cosmos/spark/CosmosClientCache.scala

```diff
@@ -31,6 +34,17 @@ private[spark] object CosmosClientCache extends BasicLoggingTrait {
   private[this] val unusedClientTtlInMs = 15 * 60 * 1000
   private[this] val cleanupIntervalInSeconds = 1 * 60
   private[this] val cache = new TrieMap[ClientConfigurationWrapper, CosmosClientCacheMetadata]
+
+  if (isOnSparkDriver) {
+    SparkSession.active.sparkContext.addSparkListener(new SparkListener() {
+      override def onApplicationEnd(applicationEnd: SparkListenerApplicationEnd) {
+        logInfo("CosmosClientCache - Spark ApplicationEnd - purging all cosmos clients");
+        cache.readOnlySnapshot().keys.foreach(clientCfgWrapper => purgeImpl(clientCfgWrapper, forceClosure = true))
```

Comment:
what about the ones in the toBeClosedWhenNotActiveAnymore?

Reply:
Good catch - yes shutdown would happen eventually...